nexus/types/src/deployment/blueprint_diff.rs

```diff
@@ -11,26 +11,621 @@ use super::blueprint_display::{
+                && !diff.blueprint_disks.added.contains_key(sled_id)
+                && !diff.blueprint_disks.modified.contains_key(sled_id)
+                && !diff.blueprint_datasets.removed.contains_key(sled_id)
+                && !diff.blueprint_datasets.added.contains_key(sled_id)
+                && !diff.blueprint_datasets.modified.contains_key(sled_id)
+            )
+        });
+
+        // Modifieds sleds are the union of sleds modified in `sled_state`,
+        // `blueprint_zones`, `blueprint_disks`, and `blueprint_datasets`.
+        //
+        // Because of backwards compatibility, disks and datasets get removed
+        // when expunged. Because of this we must also check for removed disks
+        // and datasets that are not in removed sleds.
+        let mut sleds_modified: BTreeSet<_> = diff
+            .sled_state
+            .modified
+            .keys()
+            .chain(diff.blueprint_zones.modified.keys())
+            .chain(diff.blueprint_disks.modified.keys())
+            .chain(diff.blueprint_datasets.modified.keys())
+            .map(|k| **k)
+            .collect();
+        for sled_id in diff
+            .blueprint_disks
+            .removed
+            .keys()
+            .chain(diff.blueprint_datasets.removed.keys())
+        {
+            if !sleds_removed.contains(*sled_id) {
+                sleds_modified.insert(**sled_id);
+            }
+        }
+
+        let all_sleds = sleds_added
+            .iter()
+            .chain(sleds_removed.iter())
+            .chain(sleds_modified.iter())
+            .chain(sleds_unchanged.iter())
+            .cloned()
+            .collect();
+
+        BlueprintDiffSummary {
+            before,
+            after,
+            diff,
+            all_sleds,
+            sleds_added,
+            sleds_removed,
+            sleds_modified,
+            sleds_unchanged,
+        }
+    }
+
+    /// Return a struct that can be used to display the diff.
+    pub fn display(&'a self) -> BlueprintDiffDisplay<'a> {
+        BlueprintDiffDisplay::new(self)
+    }
+
+    /// Returns whether the diff reflects any changes or if the blueprints are
+    /// equivalent.
+    pub fn has_changes(&self) -> bool {
+        // Any changes to physical disks, datasets, or zones would be reflected
+        // in `self.sleds_modified`, `self.sleds_added`, or
+        // `self.sleds_removed`.
+        if !self.sleds_modified.is_empty()
+            || !self.sleds_added.is_empty()
+            || !self.sleds_removed.is_empty()
+        {
+            return true;
+        }
+
+        self.diff.clickhouse_cluster_config.before
+            != self.diff.clickhouse_cluster_config.after
+    }
+
+    ///  The number of zones added across all sleds
+    pub fn total_zones_added(&self) -> usize {
+        self.diff
+            .blueprint_zones
+            .added
+            .values()
+            .fold(0, |acc, c| acc + c.zones.len())
```

Comment:
This is fine, but I wonder if map+sum would be clearer (albeit a line longer):

```rust
.map(|c| c.zones.len())
.sum()
```

Result should be the same, I think, although it might require `sum::<usize>()` IIRC? Up to you

Reply:
Yeah, it does require annotations, which I find distasteful. I don't find it any easier to read, but I guess if you've never used `fold` it might be.

Reply:
I left it as is.